ffi/src/transaction/transaction_id.rs

```diff
@@ -0,0 +1,182 @@
+use crate::error::ExternResult;
+use crate::handle::Handle;
+use crate::transaction::ExclusiveTransaction;
+use crate::{
+    ExternEngine, IntoExternResult, KernelStringSlice, OptionalValue, SharedExternEngine,
+    SharedSnapshot, TryFromStringSlice,
+};
+use delta_kernel::transaction::Transaction;
+use delta_kernel::{DeltaResult, Snapshot};
+use std::sync::Arc;
+
+/// Associates an app_id and version with a transaction. These will be applied to the table on commit.
+///
+/// # Returns
+/// A new handle to the transaction that will set the `app_id` version to `version` on commit
+///
+/// # Safety
+/// Caller is responsible for passing [valid][Handle#Validity] handles. CONSUMES TRANSACTION
+#[no_mangle]
+pub unsafe extern "C" fn with_transaction_id(
+    txn: Handle<ExclusiveTransaction>,
+    app_id: KernelStringSlice,
+    version: i64,
+    engine: Handle<SharedExternEngine>,
+) -> ExternResult<Handle<ExclusiveTransaction>> {
+    let txn = unsafe { txn.into_inner() };
+    let engine = unsafe { engine.as_ref() };
+    with_transaction_id_impl(*txn, app_id, version).into_extern_result(&engine)
+}
+
+fn with_transaction_id_impl(
+    txn: Transaction,
+    app_id: KernelStringSlice,
+    version: i64,
+) -> DeltaResult<Handle<ExclusiveTransaction>> {
+    let app_id_string: DeltaResult<String> = unsafe { TryFromStringSlice::try_from_slice(&app_id) };
```

Comment:
given this is `unsafe` I think we either need: `// SAFETY: ` comment on why it's actually safe or mark function as unsafe with a safety requirement in function's docstring. (I think we will have to do the latter and just say that the `app_id` passed in must be valid UTF-8, non-null ptr, etc.)

Reply:
will do, thanks @OussamaSaoudi!